### PR TITLE
8283327: Add methods to save/restore registers when calling into the VM from C1/interpreter barrier code

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that adds an API to save/restore caller-saved registers for VM upcalls for C1/interpreter.

Currently, for x86, this is done in a very ad-hoc (copy&pasty) way, which starts to fall apart. Additionally this fixes some problems with wrong stack alignment.

There is some cleanup to do separately to remove that copy&paste code in another CR. At the moment the API (`push_call_clobbered_registers/pop_call_clobbered_registers`) is only used for g1.

It's based on `RegSet` from AArch64.

Testing: tier1-5, tier1 testing with x64 and x86, with some at this point obscure combinations (like x86 UseSSE=0/1).

Thanks,
  Thomas

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283327](https://bugs.openjdk.java.net/browse/JDK-8283327): Add methods to save/restore registers when calling into the VM from C1/interpreter barrier code

### Reviewers
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to 1be201b13ea3d45ce3f55536b617e75464af9415
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to 1be201b13ea3d45ce3f55536b617e75464af9415

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7867/head:pull/7867` \
`$ git checkout pull/7867`

Update a local copy of the PR: \
`$ git checkout pull/7867` \
`$ git pull https://git.openjdk.java.net/jdk pull/7867/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7867`

View PR using the GUI difftool: \
`$ git pr show -t 7867`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7867.diff">https://git.openjdk.java.net/jdk/pull/7867.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 1]
```diff
@@ -3578,6 +3579,190 @@ void MacroAssembler::tlab_allocate(Register thread, Register obj,
...
+#else
+  regs += RegSet::of(rax, rcx, rdx);
+#endif
+  return regs;
+}
+
+XMMRegSet MacroAssembler::call_clobbered_xmm_registers() {
+#if defined(WINDOWS) && defined(_LP64)
+  XMMRegSet result = XMMRegSet::range(xmm0, xmm5);
+  if (FrameMap::get_num_caller_save_xmms() > 16) {
```
Comment (fisk) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:3599]: Why is this part needed?

Comment (tschatzl) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:3599]: In the x64 Windows ABI xmm0-5 and xmm16-31 are volatile, i.e. caller save. That's why, unlike in the SysV ABI, xmm6-15 are not in the set, and hence not saved by the caller.

The code says: if we have more than 16 xmm registers (AVX), then add xmm16 to 31 to the set of caller saved registers.
If we added xmm16 to 31 unconditionally, the code generator would complain that xmm16 to 31 are not available with e.g. SSE2 only.

Thomas

Comment (tschatzl) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:3599]: If you look below for the SysV ABI, in the '#else' part, the code unconditionally adds all xmm registers that are available:

```
XMMRegSet::range(xmm0, as_XMMRegister(FrameMap::get_num_caller_save_xmms() - 1));
```
If you want we can do the same for Win64 if you think this is better.

Comment (tschatzl) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:3599]: Of course, if C1 never uses xmm >= 8, then I can limit the number of register saving.

Comment (fisk) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:3599]: Pretty sure C1 will use more than that.

**Review by fisk — APPROVED** [review]
Looks good.

File: src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp [hunk 2]
```diff
@@ -328,15 +317,14 @@ void G1BarrierSetAssembler::g1_write_barrier_post(MacroAssembler* masm,
...
   __ bind(runtime);
   // save the live input values
-  __ push(store_addr);
+  RegSet saved = RegSet::of(store_addr NOT_LP64(COMMA thread));
+  __ push_set(saved);
 #ifdef _LP64
   __ call_VM_leaf(CAST_FROM_FN_PTR(address, G1BarrierSetRuntime::write_ref_field_post_entry), card_addr, r15_thread);
 #else
-  __ push(thread);
   __ call_VM_leaf(CAST_FROM_FN_PTR(address, G1BarrierSetRuntime::write_ref_field_post_entry), card_addr, thread);
```
Comment (dean-long) [inline src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp:325]: It looks like this #else part will work for the _LP64 case, so we don't need the #ifdef anymore.

**Review by dean-long — APPROVED** [review]
The interpreter probably doesn't need to preserve more than 1 floating pointer register.  Do we care about optimizing for that case?

**Comment by tschatzl** [comment]
> The interpreter probably doesn't need to preserve more than 1 floating pointer register. Do we care about optimizing for that case?

I can look into this as an extra CR. Not sure if it will help a lot.

**Comment by tschatzl** [comment]
Thanks @dean-long @fisk for your reviews.

/integrate

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 3]
```diff
@@ -26,6 +26,7 @@
 #include "jvm.h"
 #include "asm/assembler.hpp"
 #include "asm/assembler.inline.hpp"
+#include "c1/c1_FrameMap.hpp"
```
Comment (DamonFool) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:29]: Hi @tschatzl , this line breaks the build of VM with `--with-jvm-features=-compiler1`.
Please have a look.
Thanks.

Comment (tschatzl) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:29]: will do. Thanks!
